awscli/customizations/codedeploy/push.py

```diff
@@ -0,0 +1,287 @@
+            '--deployment-config-name <deployment-config-name> '
+            '--description <description>\n'.format(
+                params.application_name,
+                s3location_string
+            )
+        )
+
+    @contextlib.contextmanager
+    def _compress(self, source, ignore_hidden_files=False):
+        source_path = os.path.abspath(source)
+        appspec_path = os.path.sep.join([source_path, 'appspec.yml'])
+        with tempfile.TemporaryFile('w+b') as tf:
+            zf = zipfile.ZipFile(tf, 'w')
+            # Using 'try'/'finally' instead of 'with' statement since ZipFile
+            # does not have support context manager in Python 2.6.
+            try:
+                contains_appspec = False
+                for root, dirs, files in os.walk(source, topdown=True):
+                    if ignore_hidden_files:
+                        files = [fn for fn in files if not fn.startswith('.')]
+                        dirs[:] = [dn for dn in dirs if not dn.startswith('.')]
+                    for fn in files:
+                        filename = os.path.join(root, fn)
+                        filename = os.path.abspath(filename)
+                        arcname = filename[len(source_path) + 1:]
+                        if filename == appspec_path:
+                            contains_appspec = True
+                        zf.write(filename, arcname)
+                if not contains_appspec:
+                    raise RuntimeError(
+                        '{0} was not found'.format(appspec_path)
+                    )
+            finally:
+                zf.close()
+            yield tf
+
+    def _upload_to_s3(self, params, bundle):
+        size_remaining = self._bundle_size(bundle)
+        if size_remaining < MULTIPART_LIMIT:
+            return self.s3.put_object(
+                Bucket=params.bucket,
+                Key=params.key,
+                Body=bundle
+            )
+        else:
+            return self._multipart_upload_to_s3(
+                params,
+                bundle,
+                size_remaining
+            )
+
+    def _bundle_size(self, bundle):
+        bundle.seek(0, 2)
+        size = bundle.tell()
+        bundle.seek(0)
+        return size
+
+    def _multipart_upload_to_s3(self, params, bundle, size_remaining):
+        create_response = self.s3.create_multipart_upload(
+            Bucket=params.bucket,
+            Key=params.key
+        )
+        upload_id = create_response['UploadId']
+        try:
+            part_num = 1
+            multipart_list = []
+            bundle.seek(0)
+            while size_remaining > 0:
+                data = bundle.read(MULTIPART_LIMIT)
+                upload_response = self.s3.upload_part(
+                    Bucket=params.bucket,
+                    Key=params.key,
+                    UploadId=upload_id,
+                    PartNumber=part_num,
+                    Body=six.BytesIO(data)
```

Comment:
I think it was just to be safe that we were sending it as bytes. I did not really touch this part of the because I was hesitant to make changes that were not related to clients for the sake of avoiding potential regressions.